File: python/examples/client_example.py
#!/usr/bin/env python3
"""
Stratum v2 Client Example

This example demonstrates how to create a TCP client that:
1. Connects to a Stratum v2 server
2. Performs Noise_NX handshake as an initiator
3. Sends a SetupConnection message
4. Receives and parses the response

Run the server_example.py first, then run this client to see the complete flow.
"""

import socket
import base58
import time
from typing import Optional

from sv2 import (
    Sv2CodecState,
    Sv2Decoder,
    Sv2Encoder,
    Sv2Message,
    SetupConnection,
    Sv2CodecError,
    Sv2MessageError,
    SetupConnectionSuccess
)

def get_authority_public_key() -> bytes:
    """Get authority public key for connecting to the server."""
    # Use the same authority public key as the server
    authority_pub_key_b58 = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
    
    # Decode from base58 to bytes and extract 32-byte key according to SV2 spec
    pub_key_full = base58.b58decode(authority_pub_key_b58)
    authority_pub_key = pub_key_full[2:34]   # Skip 2-byte version prefix
    
    return authority_pub_key

def perform_handshake(client_socket: socket.socket, initiator: Sv2CodecState) -> bool:
    """
    Perform the 3-step Noise_NX handshake as initiator.
    
    Returns True if handshake successful, False otherwise.
    """
    try:
        print("--- Starting Handshake as Initiator ---")
        
        # Step 0: Send initiator's ephemeral public key
        print("Step 0: Sending ephemeral public key...")
        step_0_frame = initiator.step_0()
        client_socket.send(step_0_frame)
        
        print(f"✓ Sent step 0 frame: {len(step_0_frame)} bytes")
        print(f"  Ephemeral key: {step_0_frame.hex()[:32]}...")
        
        # Step 1: Receive responder's response
        print("Step 1: Waiting for responder's response...")
        step_1_data = client_socket.recv(4096)  # Receive responder's frame
        
        if not step_1_data:
            print("✗ No response received from responder")
            return False
            
        print(f"✓ Received step 1 response: {len(step_1_data)} bytes")
        print(f"  Response includes: ephemeral key + encrypted static key + signature")
        
        # Step 2: Complete handshake
        print("Step 2: Completing handshake...")
        initiator.step_2(step_1_data)
        
        print("✓ Handshake completed successfully")
        print("✓ Secure channel established")
        
        return True
        
    except Exception as e:  # Catch all exceptions since Sv2CodecError might be a subclass
        # Check if it's a codec error by checking the type name
        if 'Sv2CodecError' in str(type(e)):
            print(f"✗ Handshake failed with codec error: {e}")
        else:
            print(f"✗ Handshake failed with error: {e}")
        return False

def create_setup_connection_message() -> Optional[Sv2Message]:
    """Create a SetupConnection message to send to the server."""
    try:
        print("\n--- Creating SetupConnection Message ---")
        
        setup_connection = SetupConnection(
            protocol=1,           # Mining protocol
            min_version=2,        # Minimum protocol version  
            max_version=2,        # Maximum protocol version
            flags=0,              # Protocol flags
            endpoint_host="client.example.com",
            endpoint_port=0,      # Client doesn't listen
            vendor="Example Python Client",
            hardware_version="v1.0.0",
            firmware="py-client-1.0",
            device_id="python-client-001"
        )
        
        print("✓ Created SetupConnection message:")
        print(f"  - Protocol: {setup_connection.protocol}")
        print(f"  - Version range: {setup_connection.min_version}-{setup_connection.max_version}")
        print(f"  - Flags: {setup_connection.flags}")
        print(f"  - Endpoint: {setup_connection.endpoint_host}:{setup_connection.endpoint_port}")
        print(f"  - Vendor: {setup_connection.vendor}")
        print(f"  - Hardware Version: {setup_connection.hardware_version}")
        print(f"  - Firmware: {setup_connection.firmware}")
        print(f"  - Device ID: {setup_connection.device_id}")
        
        # Wrap in Sv2Message enum
        return Sv2Message.SETUP_CONNECTION(setup_connection)  # type: ignore
        
    except Exception as e:  # Catch all exceptions since Sv2MessageError might be a subclass
        # Check if it's a message error by checking the type name
        if 'Sv2MessageError' in str(type(e)):
            print(f"✗ Failed to create SetupConnection message: {e}")
        else:
            print(f"✗ Failed to create SetupConnection message: {e}")
        return None

def send_and_receive_messages(client_socket: socket.socket, initiator: Sv2CodecState):
    """
    Send a SetupConnection message to the server and handle the response.
    """
    encoder = Sv2Encoder()
    decoder = Sv2Decoder()
    print("\n--- Message Exchange Phase ---")
    
    # Buffer to accumulate incoming data
    data_buffer = bytearray()
    
    try:
        # Create and send SetupConnection message
        setup_message = create_setup_connection_message()
        if not setup_message:
            print("✗ Failed to create SetupConnection message")
            return
        
        print("\n--- Encoding and Sending Message ---")
        try:
            encoded_frame = encoder.encode(setup_message, initiator)  # type: ignore
            print(f"✓ Message encoded successfully: {len(encoded_frame)} bytes")
            print(f"  Encoded frame: {encoded_frame.hex()[:64]}...")
            
            client_socket.send(encoded_frame)
            print("✓ SetupConnection message sent to server")
            
        except Exception as e:
            print(f"✗ Failed to encode/send message: {e}")
            return
        
        print("\n--- Listening for Server Responses ---")
        message_count = 0
        
        while True:
            try:
                # Get the size of buffer that needs to be filled
                buffer_size = decoder.buffer_size()
                
                if buffer_size > 0:
                    # Read exactly the number of bytes the decoder needs
                    data = client_socket.recv(buffer_size)
                    
                    if not data:
                        print("✗ Server closed the connection")
                        break

                    if len(data) != buffer_size:
                        # For TCP, we might get partial data, so we need to keep reading
                        while len(data) < buffer_size:
                            more_data = client_socket.recv(buffer_size - len(data))
                            if not more_data:
                                print("✗ Server closed connection while reading")
                                return
                            data += more_data

                    # Try to decode with the exact amount of data
                    try:
                        decoded_response = decoder.try_decode(data, initiator)

                        # Successfully decoded a message!
                        message_count += 1
                        print(f"\n--- Response #{message_count} Decoded ---")

                        # Handle different response types
                        if decoded_response.is_SETUP_CONNECTION_SUCCESS():
                            print("\n🎉 Received SetupConnectionSuccess!")
                            
                            # Extract the response data
                            success_response = decoded_response[0]  # type: ignore
                            
                            print("--- SetupConnectionSuccess Details ---")
                            print(f"Used Version: {success_response.used_version}")
                            print(f"Flags: {success_response.flags}")
                            
                            print("\n✅ Connection setup completed successfully!")
                            print("Client-server communication established")
                            
                            # After successful setup, we could send more messages or exit
                            # For this example, we'll exit after receiving the success response
                            print("🏁 Example completed - connection established")
                            return
                            
                        else:
                            print(f"📨 Received message type: {type(decoded_response).__name__}")
                            
                    except Exception as e:
                        # Check if it's a MissingBytes error
                        error_type = type(e).__name__
                        
                        # Handle MissingBytes error
                        if "MissingBytes" in error_type:
                            # Decoder needs more data, will check buffer_size again
                            continue
                        else:
                            print(f"✗ Unexpected decoding error: {e}")
                            break
                else:
                    # If buffer_size is 0, try calling try_decode with empty data to trigger buffer_size calculation
                    try:
                        decoded_response = decoder.try_decode(bytes(), initiator)
                        # If this succeeds, we have a message (shouldn't happen on first call)
                        message_count += 1
                        print(f"\n--- Response #{message_count} Decoded (unexpected) ---")
                        # Handle the message as above...
                    except Exception as e:
                        # Check if it's a MissingBytes error
                        error_type = type(e).__name__
                        
                        # Handle MissingBytes error
                        if "MissingBytes" in error_type:
                            # Decoder updated buffer size, will check buffer_size again
                            continue
                        else:
                            print(f"✗ Unexpected error on initial decode: {e}")
                            break
                        
            except Exception as e:
                print(f"⚠ Error in response handling: {e}")
                break
            
    except Exception as e:
        print(f"✗ Error in message exchange: {e}")
        
    finally:
        print(f"📊 Total responses received: {message_count}")

def connect_to_server(host: str = "127.0.0.1", port: int = 34254) -> bool:
    """
    Connect to the Stratum v2 server and perform complete communication flow.
    """
    print(f"🔗 Connecting to Stratum v2 server at {host}:{port}")
    
    try:
        # Create socket and connect
        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        client_socket.settimeout(10)  # 10 second timeout
        
        print(f"Attempting to connect to {host}:{port}...")
        client_socket.connect((host, port))
        print(f"✓ Connected to server at {host}:{port}")
        
        # Get authority public key
        authority_pub_key = get_authority_public_key()
        print(f"✓ Using authority public key: {authority_pub_key.hex()[:16]}...")
        
        # Create initiator
        initiator = Sv2CodecState.new_initiator(authority_pub_key)
        print("✓ Initiator created successfully")
        
        # Perform handshake
        if perform_handshake(client_socket, initiator):
            # Send messages and handle responses after successful handshake
            send_and_receive_messages(client_socket, initiator)
        else:
            print("✗ Handshake failed, closing connection")
            return False
        
        # Keep connection alive for a moment
        print("\n--- Connection Complete ---")
        print("Keeping connection alive for 2 seconds...")
        time.sleep(2)
        
        client_socket.close()
        print("✓ Connection closed gracefully")
        return True
        
    except socket.timeout:
        print("✗ Connection timeout")
        return False
    except socket.error as e:
        if hasattr(e, 'errno') and e.errno == 61:  # Connection refused
            print(f"✗ Connection refused. Is the server running on {host}:{port}?")
            print("  Try running: python server_example.py")
        else:
            print(f"✗ Socket error: {e}")
        return False
    except Exception as e:
        print(f"✗ Connection failed: {e}")
        return False
    finally:
        try:
            client_socket.close()
        except:
            pass

def main():
    """
    Main function to demonstrate client functionality.
    """
    print("=" * 60)
    print("        Stratum v2 Client Example")
    print("=" * 60)
    print()
    print("This example demonstrates a complete Stratum v2 client that:")
    print("1. Connects to a Stratum v2 server via TCP")
    print("2. Performs Noise_NX handshake as initiator")
    print("3. Sends SetupConnection message")
    print("4. Receives SetupConnectionSuccess response")
    print()
    print("Prerequisites:")
    print("- Run the server_example.py first")
    print("- Server should be listening on 127.0.0.1:34254")
    print()
    
    try:
        # Test the connection
        success = connect_to_server()
        
        if success:
            print("\n🎉 Client example completed successfully!")
            print("\nThe client successfully:")
            print("✓ Connected to the server")
            print("✓ Completed Noise_NX handshake")
            print("✓ Sent SetupConnection message")
            print("✓ Received SetupConnectionSuccess response")
            print("\nFull Stratum v2 communication flow demonstrated!")
        else:
            print("\n❌ Client example failed")
            print("\nTroubleshooting:")
            print("1. Make sure server_example.py is running")
            print("2. Check that port 34254 is available")
            print("3. Verify network connectivity")
            
    except KeyboardInterrupt:
        print("\n\n⚠ Client interrupted by user")
    except Exception as e:
        print(f"\n✗ Unexpected error: {e}")

if __name__ == "__main__":
    main() 

File: cpp/examples/client_example.cpp
/*
 * Stratum v2 Client Example
 *
 * This example demonstrates how to create a TCP client that:
 * 1. Connects to a Stratum v2 server
 * 2. Performs Noise_NX handshake as an initiator
 * 3. Sends a SetupConnection message
 * 4. Receives and parses the response
 *
 * Run server_example.cpp first, then run this client to see the complete flow.
 */

#include <chrono>
#include <cstdint>
#include <exception>
#include <iostream>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <thread>
#include <variant>
#include <vector>

#include "example_utils.hpp"
#include "tcp_utils.hpp"
#include "sv2.hpp"

namespace {

constexpr const char *DEFAULT_HOST = "127.0.0.1";
constexpr std::uint16_t DEFAULT_PORT = 34254;
constexpr std::size_t HANDSHAKE_RESPONSE_BUFFER_SIZE = 4096;

using sv2cpp::example::tcp::TcpSocket;
using sv2cpp::example::tcp::connect_tcp;
using sv2cpp::example::tcp::recv_exact;
using sv2cpp::example::tcp::recv_some;
using sv2cpp::example::tcp::send_all;

bool perform_handshake(
    TcpSocket &client_socket,
    const std::shared_ptr<sv2::Sv2CodecState> &initiator
) {
    try {
        std::cout << "--- Starting Handshake as Initiator ---" << std::endl;

        std::cout << "Step 0: Sending ephemeral public key..." << std::endl;
        const auto step_0_frame = initiator->step_0();
        send_all(client_socket.fd(), step_0_frame);

        std::cout << "✓ Sent step 0 frame: " << step_0_frame.size() << " bytes" << std::endl;
        std::cout << "  Ephemeral key: "
                  << sv2cpp::example::shortened_hex(step_0_frame)
                  << std::endl;

        std::cout << "Step 1: Waiting for responder's response..." << std::endl;
        const auto step_1_data = recv_some(
            client_socket.fd(),
            HANDSHAKE_RESPONSE_BUFFER_SIZE
        );

        std::cout << "✓ Received step 1 response: " << step_1_data.size() << " bytes" << std::endl;
        std::cout << "  Response includes: ephemeral key + encrypted static key + signature" << std::endl;

        std::cout << "Step 2: Completing handshake..." << std::endl;
        initiator->step_2(step_1_data);

        std::cout << "✓ Handshake completed successfully" << std::endl;
        std::cout << "✓ Secure channel established" << std::endl;

        return true;
    } catch (const sv2::Sv2CodecError &error) {
        std::cout << "✗ Handshake failed with codec error: " << error.what() << std::endl;
        return false;
    } catch (const std::exception &error) {
        std::cout << "✗ Handshake failed with error: " << error.what() << std::endl;
        return false;
    }
}

std::optional<sv2::Sv2Message> create_setup_connection_message() {
    try {
        std::cout << std::endl;
        std::cout << "--- Creating SetupConnection Message ---" << std::endl;

        auto setup_connection = sv2::SetupConnection{
            .protocol = 1,
            .min_version = 2,
            .max_version = 2,
            .flags = 0,
            .endpoint_host = "client.example.com",
            .endpoint_port = 0,
            .vendor = "Example C++ Client",
            .hardware_version = "v1.0.0",
            .firmware = "cpp-client-1.0",
            .device_id = "cpp-client-001",
        };

        std::cout << "✓ Created SetupConnection message:" << std::endl;
        std::cout << "  - Protocol: " << static_cast<int>(setup_connection.protocol) << std::endl;
        std::cout << "  - Version range: "
                  << setup_connection.min_version
                  << "-"
                  << setup_connection.max_version
                  << std::endl;
        std::cout << "  - Flags: " << setup_connection.flags << std::endl;
        std::cout << "  - Endpoint: "
                  << setup_connection.endpoint_host
                  << ":"
                  << setup_connection.endpoint_port
                  << std::endl;
        std::cout << "  - Vendor: " << setup_connection.vendor << std::endl;
        std::cout << "  - Hardware Version: " << setup_connection.hardware_version << std::endl;
        std::cout << "  - Firmware: " << setup_connection.firmware << std::endl;
        std::cout << "  - Device ID: " << setup_connection.device_id << std::endl;

        return sv2::Sv2Message(
            sv2::Sv2Message::kSetupConnection{
                .message = setup_connection,
            }
        );
    } catch (const sv2::Sv2MessageError &error) {
        std::cout << "✗ Failed to create SetupConnection message: " << error.what() << std::endl;
        return std::nullopt;
    } catch (const std::exception &error) {
        std::cout << "✗ Failed to create SetupConnection message: " << error.what() << std::endl;
        return std::nullopt;
    }
}

std::optional<sv2::Sv2Message> decode_next_message(
    TcpSocket &client_socket,
    sv2::Sv2Decoder &decoder,
    const std::shared_ptr<sv2::Sv2CodecState> &initiator
) {
    while (true) {
        const auto buffer_size = decoder.buffer_size();

        if (buffer_size == 0) {
            try {
                return decoder.try_decode({}, initiator);
            } catch (const std::exception &error) {
                if (sv2cpp::example::is_missing_bytes_error(error)) {
                    continue;
                }

                throw;
            }
        }

        const auto data = recv_exact(client_socket.fd(), buffer_size);

        try {
            return decoder.try_decode(data, initiator);
        } catch (const std::exception &error) {
            if (sv2cpp::example::is_missing_bytes_error(error)) {
                continue;
            }

            throw;
        }
    }
}

void send_and_receive_messages(
    TcpSocket &client_socket,
    const std::shared_ptr<sv2::Sv2CodecState> &initiator
) {
    auto encoder = sv2::Sv2Encoder::init();
    auto decoder = sv2::Sv2Decoder::init();

    if (encoder == nullptr) {
        throw std::runtime_error("Sv2Encoder::init returned null");
    }
    if (decoder == nullptr) {
        throw std::runtime_error("Sv2Decoder::init returned null");
    }

    std::cout << std::endl;
    std::cout << "--- Message Exchange Phase ---" << std::endl;

    std::size_t message_count = 0;

    try {
        const auto setup_message = create_setup_connection_message();

        if (!setup_message.has_value()) {
            std::cout << "✗ Failed to create SetupConnection message" << std::endl;
            return;
        }

        std::cout << std::endl;
        std::cout << "--- Encoding and Sending Message ---" << std::endl;

        const auto encoded_frame = encoder->encode(*setup_message, initiator);

        std::cout << "✓ Message encoded successfully: " << encoded_frame.size() << " bytes" << std::endl;
        std::cout << "  Encoded frame: "
                  << sv2cpp::example::shortened_hex(encoded_frame, 32)
                  << std::endl;

        send_all(client_socket.fd(), encoded_frame);
        std::cout << "✓ SetupConnection message sent to server" << std::endl;

        std::cout << std::endl;
        std::cout << "--- Listening for Server Responses ---" << std::endl;

        while (true) {
            const auto decoded_response = decode_next_message(
                client_socket,
                *decoder,
                initiator
            );

            if (!decoded_response.has_value()) {
                continue;
            }

            message_count += 1;

            std::cout << std::endl;
            std::cout << "--- Response #" << message_count << " Decoded ---" << std::endl;

            const auto &variant = decoded_response->get_variant();

            if (std::holds_alternative<sv2::Sv2Message::kSetupConnectionSuccess>(variant)) {
                std::cout << std::endl;
                std::cout << "🎉 Received SetupConnectionSuccess!" << std::endl;

                const auto &success_response =
                    std::get<sv2::Sv2Message::kSetupConnectionSuccess>(variant).message;

                std::cout << "--- SetupConnectionSuccess Details ---" << std::endl;
                std::cout << "Used Version: " << success_response.used_version << std::endl;
                std::cout << "Flags: " << success_response.flags << std::endl;

                std::cout << std::endl;
                std::cout << "✅ Connection setup completed successfully!" << std::endl;
                std::cout << "Client-server communication established" << std::endl;
                std::cout << "🏁 Example completed - connection established" << std::endl;
                return;
            }

            std::cout
                << "📨 Received message type: "
                << sv2cpp::example::message_type_name(*decoded_response)
                << std::endl;
        }
    } catch (const std::exception &error) {
        std::cout << "✗ Error in message exchange: " << error.what() << std::endl;
    }

    std::cout << "📊 Total responses received: " << message_count << std::endl;
}

bool connect_to_server(
    const std::string &host = DEFAULT_HOST,
    std::uint16_t port = DEFAULT_PORT
) {
    std::cout << "🔗 Connecting to Stratum v2 server at " << host << ":" << port << std::endl;

    try {
        std::cout << "Attempting to connect to " << host << ":" << port << "..." << std::endl;
        auto client_socket = connect_tcp(host, port);
        std::cout << "✓ Connected to server at " << host << ":" << port << std::endl;

        const auto &authority_pub_key = sv2cpp::example::authority_public_key();
        std::cout << "✓ Using authority public key: "
                  << sv2cpp::example::shortened_hex(authority_pub_key, 8)
                  << std::endl;

        auto initiator = sv2::Sv2CodecState::new_initiator(authority_pub_key);
        if (initiator == nullptr) {
            throw std::runtime_error("Sv2CodecState::new_initiator returned null");
        }

        std::cout << "✓ Initiator created successfully" << std::endl;

        if (perform_handshake(client_socket, initiator)) {
            send_and_receive_messages(client_socket, initiator);
        } else {
            std::cout << "✗ Handshake failed, closing connection" << std::endl;
            return false;
        }

        std::cout << std::endl;
        std::cout << "--- Connection Complete ---" << std::endl;
        std::cout << "Keeping connection alive for 2 seconds..." << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(2));

        client_socket.close();
        std::cout << "✓ Connection closed gracefully" << std::endl;

        return true;
    } catch (const std::exception &error) {
        std::cout << "✗ Connection failed: " << error.what() << std::endl;
        std::cout << "  Is the server running on " << host << ":" << port << "?" << std::endl;
        std::cout << "  Try running: ./build/sv2cpp_example_server_example" << std::endl;
        return false;
    }
}

} // namespace

int main(int argc, char **argv) {
    std::cout << "============================================================" << std::endl;
    std::cout << "        Stratum v2 Client Example" << std::endl;
    std::cout << "============================================================" << std::endl;
    std::cout << std::endl;
    std::cout << "This example demonstrates a complete Stratum v2 client that:" << std::endl;
    std::cout << "1. Connects to a Stratum v2 server via TCP" << std::endl;
    std::cout << "2. Performs Noise_NX handshake as initiator" << std::endl;
    std::cout << "3. Sends SetupConnection message" << std::endl;
    std::cout << "4. Receives SetupConnectionSuccess response" << std::endl;
    std::cout << std::endl;
    std::cout << "Prerequisites:" << std::endl;
    std::cout << "- Run server_example.cpp first" << std::endl;
    std::cout << "- Server should be listening on 127.0.0.1:34254" << std::endl;
    std::cout << std::endl;

    const std::string host = argc >= 2 ? argv[1] : DEFAULT_HOST;
    const auto port = argc >= 3
        ? static_cast<std::uint16_t>(std::stoul(argv[2]))
        : DEFAULT_PORT;

    try {
        const auto success = connect_to_server(host, port);

        if (success) {
            std::cout << std::endl;
            std::cout << "🎉 Client example completed successfully!" << std::endl;
            std::cout << std::endl;
            std::cout << "The client successfully:" << std::endl;
            std::cout << "✓ Connected to the server" << std::endl;
            std::cout << "✓ Completed Noise_NX handshake" << std::endl;
            std::cout << "✓ Sent SetupConnection message" << std::endl;
            std::cout << "✓ Received SetupConnectionSuccess response" << std::endl;
            std::cout << std::endl;
            std::cout << "Full Stratum v2 communication flow demonstrated!" << std::endl;
            return 0;
        }

        std::cout << std::endl;
        std::cout << "❌ Client example failed" << std::endl;
        std::cout << std::endl;
        std::cout << "Troubleshooting:" << std::endl;
        std::cout << "1. Make sure server_example.cpp is running" << std::endl;
        std::cout << "2. Check that port 34254 is available" << std::endl;
        std::cout << "3. Verify network connectivity" << std::endl;

        return 1;
    } catch (const std::exception &error) {
        std::cout << std::endl;
        std::cout << "✗ Unexpected error: " << error.what() << std::endl;
        return 1;
    } catch (...) {
        std::cout << std::endl;
        std::cout << "✗ Unexpected unknown error" << std::endl;
        return 1;
    }
}